File: src/answers/typescript/easy/N_aryTreePreOrderTraversal.ts
export function preorder(root: _Node | null): number[] {
    const preOrderList: number[] = [];
    function dfs(n: _Node | null) {
        if (!n) return;
        preOrderList.push(n.val);
        for (const c of n.children) {
            dfs(c);
        }
    }
    dfs(root);
    return preOrderList;
};

class _Node {
    val: number
    children: _Node[]

    constructor(val?: number, children?: _Node[]) {
        this.val = (val === undefined ? 0 : val)
        this.children = (children === undefined ? [] : children)
    }
}

File: src/answers/python/easy/N_aryTreePreOrderTraversal.py
from typing import List, Optional

class N_aryTreePreOrderTraversal:
    def preorder(self, root: 'Node') -> List[int]:
        preOrderList = []
        def dfs(node: 'Node'):
            if not node:
                return []
            preOrderList.append(node.val)
            for c in node.children: # type: ignore
                dfs(c)
        dfs(root)        
        return preOrderList
        

class Node:
    def __init__(self, val: Optional[int] = None, children: Optional[List['Node']] = None):
        self.val = val
        self.children = children